### PR TITLE
Currency preference picker

### PR DESCRIPTION
**Pull request by cdolm92**
# What

UI for currency preference picker

# Why

Users can pick a preferred currency

# See 👀
currency-picker

# TODO ⏰

- [x] Snapshot Tests
- [x] View Model Tests
- [ ] Review Feedback

### CODE REVIEW

File: Kickstarter-iOS/Views/Cells/SettingsTableViewCell.swift [hunk 1]
```diff
@@ -1,9 +1,11 @@
 import Library
 import Prelude
+import UIKit
 
 final class SettingsTableViewCell: UITableViewCell, ValueCell, NibLoading {
+
   @IBOutlet fileprivate weak var arrowImageView: UIImageView!
-  @IBOutlet fileprivate weak var detailLabel: UILabel!
+  @IBOutlet public weak var detailLabel: UILabel!
```
Comment (cdolm92) [inline Kickstarter-iOS/Views/Cells/SettingsTableViewCell.swift]: We don't want this here. 🤔

File: Kickstarter-iOS/Views/Controllers/SettingsAccountViewController.swift [hunk 2]
```diff
@@ -78,4 +116,28 @@ extension SettingsAccountViewController: UITableViewDelegate {
...
+  }
+}
+
+extension SettingsAccountViewController: SettingsAccountPickerCellDelegate {
+  func currencyCellDetailTextUpdated(_ text: String) {
+    tableView.beginUpdates()
+    let parentIndexPath = IndexPath(row: 1, section: SettingsAccountSectionType.payment.rawValue)
+    let cell = tableView.cellForRow(at: parentIndexPath)
+    if let cell = cell as? SettingsTableViewCell {
+      cell.detailLabel.text = text // Fix this
```
Comment (cdolm92) [inline Kickstarter-iOS/Views/Controllers/SettingsAccountViewController.swift]: updating the currency cell detail text label here. needs fixing

File: Kickstarter-iOS/Views/Controllers/SettingsAccountViewController.swift [hunk 3]
```diff
@@ -53,18 +71,38 @@ final class SettingsAccountViewController: UIViewController {
...
+  func showCurrencyPickerCell(_ show: Bool) {
+    if show {
+      self.tableView.beginUpdates()
+      self.tableView.insertRows(at: [self.dataSource.insertCurrencyPickerRow()], with: .top)
+      self.tableView.endUpdates()
+    }
+  }
+
+  func dismissCurrencyPickerCell() {
+    if self.tableView.numberOfRows(inSection: SettingsAccountSectionType.payment.rawValue) == 3 {
```
Comment (cdolm92) [inline Kickstarter-iOS/Views/Controllers/SettingsAccountViewController.swift]: checking if section, `SettingsAccountSectionType.payment` is included in the section to then remove from from the section. This should be improved.

**Comment by Rcureton** [comment]
[Currency Preferences in Settings](https://trello.com/c/HbjIv05T/12-currency-preferences-in-settings)

File: Kickstarter-iOS/Views/Controllers/SettingsAccountViewController.swift [hunk 4]
```diff
@@ -23,8 +22,15 @@ final class SettingsAccountViewController: UIViewController {
...
 
     self.tableView.register(nib: .SettingsTableViewCell)
+    self.tableView.register(nib: .SettingsAccountPickerCell)
+
     self.tableView.registerHeaderFooter(nib: .SettingsHeaderView)
 
+    let tapRecognizer = UITapGestureRecognizer(target: self,
+                                               action: #selector(tapGestureToDismissCurrencyPicker))
+    tapRecognizer.cancelsTouchesInView = false
+    self.view.addGestureRecognizer(tapRecognizer)
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Controllers/SettingsAccountViewController.swift]: Is there a way we could conditionally add this gesture recognizer when the picker cell is shown, and remove it when the cell is hidden?

File: Kickstarter-iOS/Views/Cells/SettingsAccountPickerCell.swift [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
+    return Currency(rawValue: row)?.descriptionText
+  }
+
+  func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
+    guard let selectedCurrency = Currency(rawValue: row) else {
+      return
+    }
+
+    NotificationCenter.default.post(name: Notification.Name.ksr_updatedCurrencyCellDetailText,
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Cells/SettingsAccountPickerCell.swift]: There might be a way we can avoid using notifications here once we have the currency text being read from the `User` object. Can we add a `//TODO` here to remove this notification once the currency can be read from the user object?

File: Library/Currency.swift [hunk 6]
```diff
@@ -0,0 +1,72 @@
+import UIKit
+
+public enum Currency: Int {
```
Comment (ifbarrera) [inline Library/Currency.swift]: If you add `CaseIterable` protocol conformance to this enum you'll get `allCases` for free! 😃

Comment (cdolm92) [reply Library/Currency.swift]: Awesome! 🎉

File: Kickstarter-iOS/Views/Controllers/SettingsAccountViewController.swift [hunk 7]
```diff
@@ -53,18 +66,40 @@ final class SettingsAccountViewController: UIViewController {
...
+      self.tableView.insertRows(at: [self.dataSource.insertCurrencyPickerRow()], with: .top)
+      let tapRecognizer = UITapGestureRecognizer(target: self,
+                                                 action: #selector(tapGestureToDismissCurrencyPicker))
+      tapRecognizer.cancelsTouchesInView = false
+      self.view.addGestureRecognizer(tapRecognizer)
+      self.tableView.endUpdates()
+    }
+  }
+
+  func dismissCurrencyPickerCell() {
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Controllers/SettingsAccountViewController.swift]: We might need to also remove the gesture recognizer here

File: Kickstarter-iOS/Views/Controllers/SettingsAccountViewController.swift [hunk 8]
```diff
@@ -78,4 +113,18 @@ extension SettingsAccountViewController: UITableViewDelegate {
...
+  func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+    if let cell = cell as? SettingsCurrencyPickerCell {
+      cell.delegate = self
+    }
+  }
+}
+
+extension SettingsAccountViewController: SettingsCurrencyPickerCellDelegate {
+  internal func shouldDismissCurrencyPicker() {
+    tableView.beginUpdates()
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Controllers/SettingsAccountViewController.swift]: You might be able to just call `self.dismissCurrencyPickerCell()` here

**Comment by cdolm92** [comment]
g2g? 👀

**Pull request merged**